diff --git a/tlse.c b/tlse.c
index d73a981..527002a 100644
--- a/tlse.c
+++ b/tlse.c
@@ -1394,7 +1394,7 @@ static const unsigned int issurer_id[] = {1, 1, 4, 0};
 static const unsigned int owner_id[] = {1, 1, 6, 0};
 static const unsigned int validity_id[] = {1, 1, 5, 0};
 static const unsigned int algorithm_id[] = {1, 1, 3, 0};
-static const unsigned int sign_id[] = {1, 3, 2, 1, 0};
+static const unsigned int sign_id[] = {1, 3, 0};//2, 1, 0};
 static const unsigned int priv_id[] = {1, 4, 0};
 static const unsigned int priv_der_id[] = {1, 3, 1, 0};
 static const unsigned int ecc_priv_id[] = {1, 2, 0};
